sycl/doc/design/DeviceGlobal.md

```diff
@@ -563,11 +563,20 @@ instance of a device global variable in a `pi_program`.  This functionality is
 exposed as two new PI interfaces:
 
 ```
-pi_result piextCopyToDeviceVariable(pi_device Device, pi_program Program,
-  const char *name, const void *src, size_t count, size_t offset);
-
-pi_result piextCopyFromDeviceVariable(pi_device Device, pi_program Program,
-  const char *name, void *dst, size_t count, size_t offset);
+pi_result piextEnqueueDeviceVariableRead(pi_queue Queue, pi_program Program,
+                                         const char *Name, pi_bool BlockingRead,
+                                         size_t Count, size_t Offset, void *Dst,
+                                         pi_uint32 NumEventsInWaitList,
+                                         const pi_event *EventsWaitList,
+                                         pi_event *Event);
+
+pi_result piextEnqueueDeviceVariableWrite(pi_queue Queue, pi_program Program,
+                                          const char *Name,
+                                          pi_bool BlockingWrite, size_t Count,
+                                          size_t Offset, const void *Src,
+                                          pi_uint32 NumEventsInWaitList,
+                                          const pi_event *EventsWaitList,
+                                          pi_event *Event);
```

Comment:
I'm not opposed to this, but I'm curious.  Does the DPC++ runtime really need all this flexibility?  For example, will it ever use this API with `BlockingRead` set to `false`?  If not, why expose all this in the PI interface?

This is one of the criticisms I have about the PI interface in general.  It just duplicates much of the OpenCL API even if the DPC++ runtime doesn't need all that flexibility.  This makes the API more difficult to port to new backends because you need to implement stuff that is allowed in the PI API even though it's never used by the runtime.

In reality, backend ports often don't implement the PI interfaces fully because they know the runtime doesn't really needs all that functionality.  This creates a sort of double standard.  We have the "official" PI interface that no one really implements, and the "reality" version of the interface that isn't documented anywhere.

It seems like much of this would be improved if the PI interface exposed only the functionality that DPC++ really needs.  So, does the DPC++ runtime really need the extra parameters you added?

Reply:
FWIW, I don't disagree with most of your sentiments. However, the feature generally boils down to getting the pointer to the device variable and then a generic copy to or from it, so I think it makes sense to have it in line with the other memory operations, which have similar parameters to this.

Additionally, I expect to use all parameters except for `blocking_read` and `blocking_write`. Note also that the first parameter has been changed from `pi_device` to `pi_queue` as it is an enqueued operation and we can get the device from the queue but not vice versa.